_P208_Nokia_LCD_5110.ino

```diff
@@ -271,12 +277,19 @@ boolean displayText3(char deviceTemplate[][48], struct EventStruct *event ){ //4
             lcd3->setTextSize(1);
           }
           String tmpString = deviceTemplate[x];
-          String newString = "";
+          String newString = " ";
+          String newString3 = " ";
           if (tmpString.length())
           {
             newString = parseTemplate(tmpString, false);
           }else{
             newString = html_input[x]; // if webformline is empty use html input
+            // Remove trailing spaces for the next time
+            std::string newString2 (newString.c_str()); 
+            int i = newString2.find_last_not_of(" ");
+            newString2 = newString2.substr(0,i+1);
+            newString3 = newString2.c_str();
+            strncpy(html_input[x], newString3.c_str()+'\0', newString3.length()+2);

```

Comment:
This is highly inefficient and also a bit dangerous.
`c_str()` is guaranteed to be 0-terminated, so what you're doing is creating a new temporary String object, copy the data and then make a c_str() like char pointer from it.

The tricky part here is that you have no check on the max. length to copy as it may exceed the length of `html_input[x]`.

So you're copying it over and over into new objects only to copy it to a char array.
Another thing you may want to do is something like this:
`html_input[x][newString.length()] = '\0';`
This way you are sure the char array is zero-terminated. (maybe also check with the length of the char array)

Reply:
When I remove '\0', html_input[x] keeps his original length! 

I added code to maximize the content copied to html_input[x] to the length of html_input[x].

If somebody has a better piece of code to remove the trailing space's in html_input[x], please let me know.

Reply:
You can also copy first, remove spaces later.
So assume the data is already copied into `html_input`:

```c++
bool done = false;
int i = digits_per_display_line - 1;
if (newString.length() < digits_per_display_line ) {
  i = newString.length() -1;
}
for ( ; !done && i >= 0; --i) {
  const char c = html_input[x][i];
  if (c != '\0' && c != ' ') {
    done = true;
  } else {
    html_input[x][i] = '\0';
  }
}
```

Reply:
Tanks. I used a other suggestion from you. Result:

            while (newString.endsWith(" ")) {
              newString = newString.substring(0, newString.length() - 1);
            }
            int len = newString.length();
            if (len > digits_per_display_line) {len = digits_per_display_line;};  // max len is len html_input[x]
            newString = newString.substring(0,len);
            strncpy(html_input[x], newString.c_str(), len);